SmugMug v2 integration

joanniclaborde opened the pull request:
Adds [SmugMug v2 API](https://api.smugmug.com/api/v2/doc/reference/index.html) support to the photo widget.

A new optional `step` argument is also added to the `pagination` helper, which defaults to 1. This is to support resources that use offsets for pagination, like SmugMug. For example, if the current page is 12, and `step` is 3, then the previous page parameter will be 9, and the next page will be 15.

A new `filter` mode is added to the photo widget, with the `default` and `gallery` views. It's basically the old Clique Tools tag widget.

pushred commented on widgets/photo/browse.js:
```diff
@@ -1,5 +1,6 @@
-var wordpress = require('../utils/wordpress');
+var _          = require('underscore');
+var wordpress  = require('../utils/wordpress');
+var smugmug_v2 = require('../utils/smugmug_v2');
```
We're not bothering specifying the versions on the other APIs, why start doing it here?

joanniclaborde replied:
Is there any chance we'll be using another version of SmugMug?

pushred replied:
Not anytime soon, version 2 has been in development for at least a year and it's _still_ in beta hah.

pushred commented on widgets/photo/browse.js:
```diff
@@ -9,6 +10,9 @@ module.exports = function(options) {
   case 'wordpress/categories':
     options.hash.data = wordpress.normalizeCategories(this, options.hash.data);
     break;
+  case 'smugmug-v2/AlbumImages':
```
The other types are lowercased, why switch to titleized case here? `images` would be fine, they can't belong to anything other than albums/galleries.

joanniclaborde replied:
I simply followed the [API naming](https://api.smugmug.com/api/v2/album/9q72gr!images), since this string refers to the raw API data, not the normalized result which is consistent through all APIs.

pushred replied:
Yeah I just don't see the value in doing that, we're otherwise abstracting the resources away.

pushred commented on widgets/utils/smugmug_v2.js:
```diff
@@ -0,0 +1,66 @@
+
+  return {
+    url:           largest_image.Url,
+    thumbnail_url: album_image.ThumbnailUrl,
```
This will be too small generally and they're automatically cropped into a square, losing their aspect ratio. SmugMug's small size would be a better fit, accessible at `album_image.Uris.ImageSizes.ImageSizes.SmallImageUrl` with this expansion:

https://api.smugmug.com/api/v2/folder/user/rascalflatts/Official-Photos!albums?_expand=AlbumHighlightImage.ImageSizes&_expandmethod=inline

The thumbnails will likely be further resized to the exact dimensions using imgix eventually.

pushred commented on helpers/pagination.js:
```diff
@@ -1,12 +1,13 @@
 module.exports = function(data, per_page, page, options) {
   var result       = '';
   var current_page = parseFloat(page || '1', 10);
+  var step         = parseFloat(options.hash.step || '1', 10);
```
This assumes the count is 10? Most APIs provide pagination resource URLs these days, including SmugMug &mdash; [see (the start of) my docs](https://github.com/SparkartGroupInc/jungle-solidus/wiki/Pagination). What about parsing these parameter values out of these URLs, keeping the pagination easily in sync with the desired count and skipping rendering the prev/next links when there's no pages remaining?

WP-API also provides URLs, but [they're basically impossible to get to](http://wp-api.org/guides/getting-started.html#getting-posts) right now. The likelihood of running out of content to advance to is much less likely there though.

joanniclaborde replied:
My error, `parseFloat` does not take a base like `parseInt` does, I'll remove the 10 parameters.

For the pagination, yeah we could do that, each widget would be responsible to provide the previous and next page numbers, and the `pagination` helper could be dropped entirely.

pushred replied:
Oh that's the radix. I think it's just best to keep these params in the resource config.

pushred commented:
I think we're going to need pagination on the display mode, see the nav on Rascal Flatts and Janet Jackson. The whole reason `ajax-photo` exists on Rascal Flatts is to leverage the serverside pagination.

I see two options for implementing:

 - Load all images and implement simple cursoring pagination using the selected image key and finding the previous/next images

 - Load all image and apply [SmugMug's own pagination parameters](https://api.smugmug.com/api/v2/doc/tutorial/paging.html), populating `start` with the selected photo's index. This would possibly render the photo keys useless if you used that resource for rendering a selected photo. But it'd probably be best to reserve that for pagination only and continue to load the selected image separately so the keys reliably load the same resource.

joanniclaborde commented:
I addressed all your comments, except the display mode pagination. For that, I think option 1 is best too: if the widget receives an array, it finds the matching topic, which in turn gives him the previous and next topics.

If we find out the resources are too big, the browse modes could add the current `page` parameter to the display page links. The appropriate resources will need to be setup on the display page, and the browsing will be limited to the current batch of items. To fix that, we could preload the previous and next pages, etc etc etc. So I suggest we wait to see if it's a real problem, it could get complex quick :(

joanniclaborde commented:
Ok this is weird, I was wondering where the Previous and Next thumbnails on [Janet's site](http://www.janetjackson.com/photos/Live) where coming from, it turns out that page is not using the display mode at all, but the browse mode with each page size being 1. See https://cliquetools-developer.sparkart.net/developer/209/pages/273/edit/pages/photos

Do we still need display mode pagination?

pushred commented:
Ahh right you are, missed the fact that there wasn't a topic in that URL. Rascal Flatts is using the display mode pagination though.

joanniclaborde commented:
Should be all set now.

pushred commented:
:+1: Only thing is this `start=0` default madness.. any word from them?

joanniclaborde commented:
They just replied, with a ridiculous answer...

> Thanks for reporting this, we're always interested to hear feedback. Using an empty start parameter is invalid input, so you should always specify a number. I'll make a note to look into updating that in the future and provide an appropriate default.

joanniclaborde commented:
Until they fix the problem, I found a workaround: I store the whole start parameter in the page parameter:

`http://site.com/photo/all?page=start%3D10`
`https://api.smugmug.com/api/v2/album/{tag}!images?count=5&{page}`

The first page is then valid.

joanniclaborde commented:
Fixed, @pushred can you think of a problem with this?

pushred commented:
So we still need to include a parameter in all the links to the page? What's the difference between this approach and setting `page=1` on all inbound links?

joanniclaborde commented:
No other changes are necessary on the site. When no page is specified, the resource ends up looking like this `/api/v2/album/{tag}!images?count=5&` instead of `/api/v2/album/{tag}!images?count=5&start=`, which was causing the problem.

pushred commented:
Huh, so when `start` isn't present at all it defaults to 1, but when the key is there it's `0`?

joanniclaborde commented:
Right.

pushred commented:
Almost everyone ignores valueless params.. but this is SmugMug we're talking about! I can't think of any resulting problems. It's abstracted away into the widgets.
